create LSP25 Execute Relay Call standard, extracting functions from LSP6 standard (#224)
lukso-network/LIPs: LSPs/LSP-25-ExecuteRelayCall.md

    - [How nonces are represented across channels?](#how-nonces-are-represented-across-channels-)
- [Implementation](#implementation)
- [Interface Cheat Sheet](#interface-cheat-sheet)
- [Copyright](#copyright)

# Simple Summary

This standard describes an interface and a signature scheme that can be used to integrate meta-transactions (transactions where users do not pay for the gas) inside a smart contract.

# Abstract

Meta transactions in web3 are used in many different context. From interacting with Universal Profiles to transferring tokens and NFTs to voting systems, where delegates can submit transactions on behalf of other users and pay for the gas cost.

# Motivation

By having a common smart contract interface, applications and protocols can implement the LSP25 Execute Relay Call standard to start implementing gas-less transaction. This can be beneficial for protocols and applications like Digital Marketplaces or Universal Profile that want to onboard new users without requiring users to hold native tokens.

# Specification

**LSP25-ExecuteRelayCall** interface id according to [ERC165]: `0x5ac79908`

> Smart contracts implementing the LSP6 standard MUST implement the [ERC165] `supportsInterface(..)` function and MUST support the [ERC165] and LSP25 interface ids.

Reviewer: Since we are linking ERC165 would it make sense to turn LSP6 into links as well?

Reviewer: It's only about convenience.

Author: It's an error, it should be LSP25, not LSP6. Thanks for flagging!

## Methods

### getNonce

```solidity
function getNonce(address signer, uint128 channel) external view returns (uint256)
```

Returns the latest nonce for a signer on a specific channel. A signer can choose a channel number arbitrarily and use this nonce to sign a payload that can be executed as a meta-transaction by any address via [executeRelayCall](#executeRelayCall) function.

_Parameters:_

- `signer`: the address of the signer of the transaction.
- `channel` : the channel which the signer wants to use for executing the transaction.

_Returns:_ `uint256` , the current nonce.

> Payloads signed with incremental nonces on the same channel for the same signer are executed in order. e.g, in channel X, the payload nb two signed with the second nonce will not be successfully executed until the payload nb one signed with the first nonce has been executed.

Reviewer: There's a typo?
`.. payload nb two ...`

Didn't get what `nb` means.

Reviewer: I see `nb` repeating further as well.

Author: Will reword this

Payloads signed with nonces on different channels are executed independently from each other, regardless of when they got executed or if they got executed successfully or not. e.g, the payload signed with the fourth nonce on channel X can be successfully executed even if the payload signed with the first nonce of channel Y:

- was executed before.
- was executed and reverted.

> X and Y can be any arbitrary number between 0 and 2^128.

Read [what are multi-channel nonces](#what-are-multi-channel-nonces).

### executeRelayCall

```solidity
function executeRelayCall(
    bytes memory signature, 
    uint256 nonce, 
    uint256 validityTimestamps, 
    bytes memory payload
) 
    external 
    payable 
    returns (bytes memory)
```

Allows anybody to execute a `payload` given they have a valid signature from a signer. The signature MUST be formed according to the [LSP25 Signature specification format](#signature-format).

_Parameters:_

- `signature`: A 65 bytes long ethereum signature.
- `nonce`: MUST be the nonce of the address that signed the message. This can be obtained via the `getNonce(address address, uint256 channel)` function.

> - `validityTimestamps`: Two `uint128` timestamps concatenated together. The first timestamp determines from when the payload can be executed, the second timestamp determines a deadlines after which the payload is not valid anymore. If validityTimestamps is `0`, the payload is valid at indefinitely at any point in time and the checks for the timestamps are skipped.

Reviewer: ```suggestion
- `validityTimestamps`: Two `uint128` timestamps concatenated together. The first timestamp determines from when the payload can be executed, and the second timestamp determines a deadline after which the payload is no longer valid. If `validityTimestamps` is `0`, the payload is valid indefinitely at any point in time and the checks for the timestamps are skipped.
```

> - `payload`: The abi-encoded function call to be executed on the linked target contract.

Reviewer: LSP6 has specifications about the linked target contract while LSP25 says nothing about it. 
Abstract of LSP6 mentions linked target contract and LSP6 standard also documents `function target() external view returns (address)`.
Nothing here in LSP25 and again it was somewhat confusing.

Reviewer: It looks like you are mentioning here parts of LSP6 while LSP25 only `requires: ERC165`.
Since LSP6 depends on LSP25 it would make sense to reference bits and pieces of LSP25 within LSP6 but not the other way around.

It feels like `function target()` should be moved to this LSP as well.
What do you think?

Author: @Jenea There should not be `target()` in this specs. I have re-worded the paragraph. Thanks for pointing it out :pray:

> _Returns:_ `bytes` , the returned data as abi-decoded bytes of the call on ERC725 smart contract, if the call succeeded, otherwise revert with a reason-string.

Reviewer: Nothing up until this point suggested that the target contract should actually be an "ERC725 smart contract".
I'm highlighting this as this standard is generic and can be cross-chain.
